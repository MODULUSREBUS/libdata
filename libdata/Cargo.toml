[package]
name = "libdata"
version = "0.1.0"
authors = ["Martin Toman <toman.martin@live.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
datacore = { path = "../datacore" }
protocol = { path = "../protocol" }
anyhow = "1"
rand_chacha = "0.2"
blake3 = "1.3"
futures-lite = "1.12"
tokio = { version = "1.23", features = [ "sync" ] }
tokio-stream = "0.1"
async-trait = "0.1"
hex = "0.4"
multi-map = "1.3"

[dev-dependencies]
index-access-memory = { git = "https://github.com/MODULUSREBUS/index-access" }
quickcheck = "1.0"
insta = "1.22"
sluice = "0.5"
tokio = { version = "1.23", features = [ "full" ] }
async-compat = "0.2"
