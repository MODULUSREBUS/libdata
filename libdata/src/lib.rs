#![forbid(unsafe_code, bad_style, nonstandard_style, future_incompatible)]
#![forbid(rust_2018_idioms, rust_2021_compatibility)]
#![deny(missing_docs)]
#![cfg_attr(test, allow(dead_code))]
#![cfg_attr(test, deny(warnings))]

//! Libdata re-exports public interface from [datacore],
//! defines async [CoreIterator],
//! defines interface for managing collection of [Cores],
//! and specifies [replication] over [protocol].

mod cores;
mod iter;

pub mod key;
pub mod keypair;
pub mod replication;

pub use datacore::ed25519_dalek;
pub use datacore::{Core, IndexAccess, Signature, MAX_CORE_LENGTH};

pub use cores::Cores;
pub use iter::CoreIterator;
pub use key::{DiscoveryKey, PublicKey, SecretKey};
pub use keypair::Keypair;

