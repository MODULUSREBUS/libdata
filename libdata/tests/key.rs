use insta;
use quickcheck::{quickcheck, TestResult};

use libdata::{keypair, KeyPair};

#[test]
fn key_can_generate() {
    let _keypair = KeyPair::generate();
}

#[test]
fn key_can_derive() {
    let keypair = KeyPair::generate();
    let _derived = keypair::derive(&keypair.sk, "hello");
}

quickcheck! {
    fn key_same_key_different_names(a: String, b: String) -> TestResult {
        if a == b {
            return TestResult::discard()
        }

        let main = KeyPair::generate();
        let a = keypair::derive(&main.sk, &a);
        let b = keypair::derive(&main.sk, &b);

        TestResult::from_bool(a.as_slice() != b.as_slice())
    }

    fn key_different_key_same_name(name: String) -> bool {
        let a = KeyPair::generate();
        let b = KeyPair::generate();
        let a = keypair::derive(&a.sk, &name);
        let b = keypair::derive(&b.sk, &name);

        a.as_slice() != b.as_slice()
    }

    fn key_same_key_same_name(name: String) -> bool {
        let main = KeyPair::generate();
        let a = keypair::derive(&main.sk, &name);
        let b = keypair::derive(&main.sk, &name);

        a.as_slice() == b.as_slice()
    }
}

const SEED_BYTES: [u8; 32] = [
    157, 097, 177, 157, 239, 253, 090, 096, 186, 132, 074, 244, 146, 236, 044, 196, 068, 073, 197,
    105, 123, 050, 105, 025, 112, 059, 172, 003, 028, 174, 127, 096,
];

#[test]
fn key_secret_key_bytes_have_not_changed() {
    let main = KeyPair::from_seed(keypair::Seed::from(SEED_BYTES)).sk;
    insta::assert_debug_snapshot!(main.as_slice());
}

#[test]
fn key_snapshot_1() {
    let main = KeyPair::from_seed(keypair::Seed::from(SEED_BYTES)).sk;
    let keypair = keypair::derive(&main, "hello");
    insta::assert_debug_snapshot!(keypair.as_slice());
}

#[test]
fn key_snapshot_2() {
    let main = KeyPair::from_seed(keypair::Seed::from(SEED_BYTES)).sk;
    let keypair = keypair::derive(&main, "hello2");
    insta::assert_debug_snapshot!(keypair.as_slice());
}

#[test]
fn key_snapshot_3() {
    let main = KeyPair::from_seed(keypair::Seed::from(SEED_BYTES)).sk;
    let keypair = keypair::derive(
        &main,
        "a very long string as a key name should not break the key derive, \
        it should just work without any issues, this is just testing it, \
        to be sure",
    );
    insta::assert_debug_snapshot!(keypair.as_slice());
}
