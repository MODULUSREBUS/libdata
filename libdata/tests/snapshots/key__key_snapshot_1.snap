---
source: libdata/tests/key.rs
expression: keypair.as_slice()
---
[
    100,
    234,
    209,
    137,
    37,
    211,
    93,
    136,
    193,
    21,
    98,
    61,
    16,
    92,
    64,
    180,
    4,
    211,
    140,
    67,
    45,
    28,
    198,
    92,
    185,
    169,
    91,
    42,
    45,
    134,
    221,
    100,
    83,
    135,
    70,
    138,
    218,
    161,
    78,
    193,
    114,
    102,
    116,
    225,
    71,
    76,
    77,
    196,
    72,
    238,
    193,
    1,
    173,
    164,
    154,
    151,
    159,
    155,
    12,
    200,
    231,
    2,
    193,
    222,
]
